apps/src/gamelab/GameLab.js

```diff
@@ -826,38 +826,132 @@ GameLab.prototype.onP5ExecutionStarting = function () {
 };
 
 /**
- * This is called while this.gameLabP5 is in the preload phase. We initialize
- * the interpreter, start its execution, and call the user's preload function.
+ * This is called while this.gameLabP5 is in the preload phase. Do the following:
  *
- * @return {Boolean} whether or not the preload has completed
+ * - load animations into the P5 engine
+ * - initialize the interpreter
+ * - start its execution
+ * - (optional) execute global code
+ * - call the user's preload function
+ *
+ * @return {Boolean} whether or not the preload has completed.  If we return
+ *         FALSE here, P5 will internally increment a preload count; calling
+ *         notifyPreloadPhaseComplete is then necessary to continue starting
+ *         the game.
  */
 GameLab.prototype.onP5Preload = function () {
-  this.gameLabP5.preloadAnimations(this.studioApp_.reduxStore.getState().animationList);
-
-  this.initInterpreter();
-  // And execute the interpreter for the first time:
-  if (this.JSInterpreter && this.JSInterpreter.initialized()) {
-    // Start executing the interpreter's global code as long as a setup() method
-    // was provided. If not, we will skip running any interpreted code in the
-    // preload phase and wait until the setup phase.
-    this.preloadInProgress = true;
-    if (this.globalCodeRunsDuringPreload) {
-      this.JSInterpreter.executeInterpreter(true);
-      this.interpreterStarted = true;
+  let isReadyNow = false;
+  Promise.all([
+      this.preloadAnimations_(),
+      this.runPreloadEventHandler_()
+  ]).then(() => {
+    isReadyNow = true;
+    this.gameLabP5.notifyPreloadPhaseComplete();
+  });
+  return isReadyNow;
+};
 
-      // In addition, execute the global function called preload()
-      if (this.eventHandlers.preload) {
-        this.eventHandlers.preload.apply(null);
-      }
+/**
+ * Wait for animations to be loaded into memory and ready to use, then pass
+ * those animations to P5 to be loaded into the engine as animations.
+ * @returns {Promise} which resolves once animations are in memory in the redux
+ *          store and we've started loading them into P5.
+ *          Loading to P5 is also an async process but it has its own internal
+ *          effect on the P5 preloadCount, so we don't need to track it here.
+ * @private
+ */
+GameLab.prototype.preloadAnimations_ = function () {
+  let store = this.studioApp_.reduxStore;
+  return new Promise(resolve => {
+    if (this.areAnimationsReady_()) {
+      resolve();
     } else {
-      if (this.eventHandlers.preload) {
-        this.log("WARNING: preload() was ignored because setup() was not provided");
-        this.eventHandlers.preload = null;
+      // Watch store changes until all the animations are ready.
+      const unsubscribe = store.subscribe(() => {
+        if (this.areAnimationsReady_()) {
+          unsubscribe();
```

Comment:
Unfortunately yes: [Store.subscribe](http://redux.js.org/docs/api/Store.html#subscribe) returns its unsubscribe function, and it's the only way to unsubscribe.